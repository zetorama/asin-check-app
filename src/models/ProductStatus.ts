export enum ProductStatus {
    New = 'new',
    InQueue = 'inqueue',
    Refreshing = 'refreshing',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export default ProductStatus
