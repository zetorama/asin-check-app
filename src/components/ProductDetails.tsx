import React, { Component, Fragment } from 'react'
import { ListGroup, ListGroupItem, Button, Alert } from 'reactstrap'
import { observer } from 'mobx-react'

import { ProductStatus, ProductModel } from '../models/Product'
import autobind from '../utils/autobind'

const InfoItem = ({ label, children }: { label: string; children: React.ReactNode }) => (
    <ListGroupItem>
        <strong className='mr-2'>{label}:</strong>
        {children}
    </ListGroupItem>
)

const EmptyValue = () => <>n/a</>

export interface ProductProps {
    product: ProductModel
    onRefresh?: (e: React.MouseEvent, product: ProductModel) => void
}

export class ProductDetails extends Component<ProductProps> {
    render() {
        const { onRefresh, product } = this.props
        const { isLoading, asin, data, updatedDateTime } = product

        return (
            <div className='well'>
                <h1>
                    {/* explicit spacing just makes text select easier */}
                    {'ASIN '}
                    <code>{asin}</code>
                    {isLoading && ' is refreshing…'}
                </h1>
                {this.renderTitle()}
                <ListGroup className={(isLoading && 'text-muted') || undefined}>
                    {data ? this.renderDetails() : this.renderExcuse()}

                    <InfoItem label='Updated On'>
                        <span className='mr-2'>{updatedDateTime || 'Unknown'}</span>
                        {onRefresh && (
                            <Button
                                className='details-refresh'
                                outline
                                color='primary'
                                size='sm'
                                onClick={this.handleRefresh}
                            >
                                {isLoading && 'Re-init '}
                                Refresh
                            </Button>
                        )}
                    </InfoItem>
                </ListGroup>
            </div>
        )
    }

    @autobind
    handleRefresh(e: React.MouseEvent) {
        const { onRefresh, product } = this.props
        onRefresh && onRefresh(e, product)
    }

    renderTitle() {
        const { data, isLoading } = this.props.product
        if (!data || !data.title) return

        return <h2 className={(isLoading && 'text-muted') || undefined}>{data.title}</h2>
    }

    renderExcuse() {
        const { status } = this.props.product
        switch (status) {
            case ProductStatus.Timeout:
                return (
                    <ListGroupItem>
                        <Alert color='warning'>Sync is canceled due to timeout. Please, try again</Alert>
                    </ListGroupItem>
                )
            case ProductStatus.Nomatch:
                return (
                    <ListGroupItem>
                        <Alert color='warning'>Cannot find Product with this ASIN. Sorry for that</Alert>
                    </ListGroupItem>
                )
            case ProductStatus.Fail:
                return (
                    <ListGroupItem>
                        <Alert color='danger'>
                            There was some fatal error during last sync. I wish I had more information
                        </Alert>
                    </ListGroupItem>
                )
        }
    }

    renderDetails() {
        const { data } = this.props.product
        if (!data) return null

        const rating = data.rating || <EmptyValue />
        const dimensions = data.dimensions || <EmptyValue />
        const categories = data.categories && data.categories.length ? data.categories.join(' › ') : <EmptyValue />
        const ranks =
            data.rank && data.rank.length ? (
                data.rank.map((rank, index) => (
                    <Fragment key={index}>
                        <br />
                        {rank}
                    </Fragment>
                ))
            ) : (
                <EmptyValue />
            )

        return (
            <>
                <InfoItem label='Category'>{categories}</InfoItem>
                <InfoItem label='Best Sellers Rank'>{ranks}</InfoItem>
                <InfoItem label='Product Dimensions'>{dimensions}</InfoItem>
                <InfoItem label='Rating'>{rating}</InfoItem>
            </>
        )
    }
}

export default observer(ProductDetails)
