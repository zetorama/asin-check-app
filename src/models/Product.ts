import { getParent, types, Instance, SnapshotIn } from 'mobx-state-tree'

import { StoreModel } from './Store'

export enum ProductStatus {
    Init = 'init',
    Queue = 'queue',
    Fetch = 'fetch',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export const ProductDetails = types.model('ProductDetails', {
    title: types.string,
    categories: types.array(types.string),
    dimensions: types.maybe(types.string),
    rank: types.maybe(types.array(types.string)),
    rating: types.maybe(types.string),
})

export type ProductAsin = string

export type ProductModel = Instance<typeof Product>
export type ProductSnapshot = SnapshotIn<typeof Product>

export const Product = types
    .model('Product', {
        asin: types.identifier,
        status: types.enumeration<ProductStatus>('ProductStatus', Object.values(ProductStatus)),
        data: types.maybe(types.frozen(ProductDetails)),

        createdOn: types.maybe(types.string),
        updatedOn: types.maybe(types.string),
        isViewed: false,
    })
    .views((self) => ({
        get isLoading(): boolean {
            return (
                self.status === ProductStatus.Init ||
                self.status === ProductStatus.Queue ||
                self.status === ProductStatus.Fetch
            )
        },
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    }))
    .actions((self) => ({
        toggleViewed(setAsViewed: boolean = !self.isViewed) {
            self.isViewed = setAsViewed
        },
    }))

export default Product
