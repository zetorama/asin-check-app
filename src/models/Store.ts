import { types, getParent, getSnapshot, applySnapshot, Instance, SnapshotIn } from 'mobx-state-tree'

import { fetchProductsLog, saveProductsLog, fetchProductsByAsins, refreshProductsByAsins } from '../services/product'
import { Product, ProductAsin, ProductModel, ProductStatus, ProductSnapshot } from './Product'

export const PRODUCTS_LIMIT = 12

export type StoreModel = Instance<typeof Store>
export type StoreSnapshot = SnapshotIn<typeof Store>

type Void<T> = { [P in keyof T]?: undefined }

const resetObj = <T>(source: T): Void<T> =>
    Object.keys(source).reduce((all, one) => Object.assign(all, { [one]: undefined }), {})

export const ProductInStore = types.compose(
    Product,
    types.model('ProductInStore').views((self) => ({
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    })),
)

export type ProductInStoreModel = Instance<typeof ProductInStore>
export type ProductInStoreSnapshot = SnapshotIn<typeof ProductInStore>

export const Store = types
    .model('AppStore', {
        version: '0.0.0',
        products: types.array(ProductInStore),
        viewingProduct: types.maybe(types.reference(ProductInStore)),
        lookingUpValue: types.optional(types.string, ''),
    })
    .views((self) => ({
        get loadingProducts(): ProductModel[] {
            return self.products.filter((p) => p.isLoading)
        },
        getProduct(asin: ProductAsin): ProductModel | undefined {
            return self.products.find((p) => p.asin === asin)
        },
    }))
    .actions((self) => ({
        mergeUpdates(products: ProductSnapshot[]) {
            // Update only existed items, keeping the order
            for (const update of products) {
                const product = self.getProduct(update.asin)
                if (product) {
                    applySnapshot(product, Object.assign(resetObj(product), update))

                    if (product === self.viewingProduct) {
                        // Mark as seen, as its currently opened
                        product.toggleViewed(true)
                    }
                }
            }
        },
        createProduct(asin: ProductAsin): ProductInStoreModel {
            const product = self.getProduct(asin)
            const defaults = product ? getSnapshot(product) : undefined
            return ProductInStore.create({
                ...defaults,
                asin,
                status: ProductStatus.New,
            })
        },
        addProducts(products: ProductModel[]) {
            if (!products || !Array.isArray(products) || !products.length) return

            const oldProducts = self.products.filter((p1) => !products.find((p2) => p1.asin === p2.asin))
            const newProducts: ProductInStoreModel[] = products.concat(oldProducts).slice(0, PRODUCTS_LIMIT)
            self.products.replace(newProducts)
        },
        clearLookingUpValue() {
            self.lookingUpValue = ''
        },
        viewProduct(product?: ProductModel | ProductAsin) {
            self.viewingProduct = product
                ? self.getProduct(ProductInStore.is(product) ? product.asin : product)
                : undefined
            if (self.viewingProduct && !self.viewingProduct.isViewed) {
                self.viewingProduct.toggleViewed(true)
            }
        },
    }))
    .actions((self) => ({
        saveProductsLog() {
            saveProductsLog(getSnapshot(self.products))
        },
        async fetchProductsLog() {
            const products = (await fetchProductsLog()) as ProductModel[]
            products && products.length && self.addProducts(products)
        },
        async lookup(value: string) {
            self.lookingUpValue = String(value || '')

            if (self.lookingUpValue) {
                const asins = self.lookingUpValue.split(/\W+/)
                const products = asins
                    .filter((asin, index) => asin && asins.indexOf(asin) === index)
                    .map((asin) => self.createProduct(asin))

                self.viewingProduct = undefined
                await self.addProducts(products)

                // switch to a newly added product
                self.viewProduct(self.products[0])
                self.clearLookingUpValue()
            }
        },
        async fetchUpdates(asins: ProductAsin[] = self.loadingProducts.map((p) => p.asin)) {
            const update = await fetchProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
        async refreshProducts(asins: ProductAsin[]) {
            for (const asin of asins) {
                const product = self.getProduct(asin)
                product && product.setStatus(ProductStatus.InQueue)
            }
            const update = await refreshProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
    }))

export default Store
