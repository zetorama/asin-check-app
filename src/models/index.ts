import { Store, StoreModel, StoreSnapshot } from './Store'

export * from './Store'
export * from './Product'

export function createStore(initial?: StoreSnapshot): StoreModel {
    return Store.create(initial)
}

export default createStore
