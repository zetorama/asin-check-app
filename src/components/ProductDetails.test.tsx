import * as React from 'react'
import { shallow } from 'enzyme'

import { ProductDetails } from './ProductDetails'
import { Product, ProductStatus } from '../models'

describe('ProductDetails', () => {
    it('renders with empty product', () => {
        const product = Product.create({
            asin: 'XXX',
            status: ProductStatus.Fail,
        })
        const cmp = shallow(<ProductDetails product={product} />)

        expect(cmp).toMatchSnapshot()
    })

    it('renders with full product', () => {
        const product = Product.create({
            asin: 'XXX',
            status: ProductStatus.Success,
            data: {
                title: 'xxx aaa',
                categories: ['xxx', 'aaa'],
                dimensions: '1 x 2 x 3',
                rank: ['1', '2', '3'],
                rating: 'ok',
            },
        })
        const cmp = shallow(<ProductDetails product={product} />)

        expect(cmp).toMatchSnapshot()
    })

    it('renders with refresh button', () => {
        const cb = jest.fn()
        const product = Product.create({
            asin: 'XXX',
            status: ProductStatus.Success,
        })
        const cmp = shallow(<ProductDetails product={product} onRefresh={cb} />)

        expect(cmp).toMatchSnapshot()
    })

    it('calls refresh callback', () => {
        const cb = jest.fn()
        const product = Product.create({
            asin: 'XXX',
            status: ProductStatus.Success,
        })
        const cmp = shallow(<ProductDetails product={product} onRefresh={cb} />)

        cmp.find('.details-refresh').simulate('click')
        expect(cb).toHaveBeenCalledTimes(1)
        expect(cb.mock.calls[0][1]).toHaveProperty('asin', 'XXX')
    })
})
