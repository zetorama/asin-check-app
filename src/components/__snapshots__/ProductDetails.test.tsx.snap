// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails renders with empty product 1`] = `
<div
  className="well"
>
  <h1>
    ASIN 
    <code>
      XXX
    </code>
  </h1>
  <ListGroup
    tag="ul"
  >
    <ListGroupItem
      tag="li"
    >
      <Alert
        closeAriaLabel="Close"
        color="danger"
        fade={true}
        isOpen={true}
        tag="div"
        transition={
          Object {
            "appear": true,
            "baseClass": "fade",
            "baseClassActive": "show",
            "enter": true,
            "exit": true,
            "in": true,
            "mountOnEnter": false,
            "onEnter": [Function],
            "onEntered": [Function],
            "onEntering": [Function],
            "onExit": [Function],
            "onExited": [Function],
            "onExiting": [Function],
            "tag": "div",
            "timeout": 150,
            "unmountOnExit": true,
          }
        }
      >
        There was some fatal error during last sync. I wish I had more information
      </Alert>
    </ListGroupItem>
    <InfoItem
      label="Updated On"
    >
      <span
        className="mr-2"
      >
        Unknown
      </span>
    </InfoItem>
  </ListGroup>
</div>
`;

exports[`ProductDetails renders with full product 1`] = `
<div
  className="well"
>
  <h1>
    ASIN 
    <code>
      XXX
    </code>
  </h1>
  <h2>
    xxx aaa
  </h2>
  <ListGroup
    tag="ul"
  >
    <InfoItem
      label="Category"
    >
      xxx â€º aaa
    </InfoItem>
    <InfoItem
      label="Best Sellers Rank"
    >
      <br />
      1
      <br />
      2
      <br />
      3
    </InfoItem>
    <InfoItem
      label="Product Dimensions"
    >
      1 x 2 x 3
    </InfoItem>
    <InfoItem
      label="Rating"
    >
      ok
    </InfoItem>
    <InfoItem
      label="Updated On"
    >
      <span
        className="mr-2"
      >
        Unknown
      </span>
    </InfoItem>
  </ListGroup>
</div>
`;

exports[`ProductDetails renders with refresh button 1`] = `
<div
  className="well"
>
  <h1>
    ASIN 
    <code>
      XXX
    </code>
  </h1>
  <ListGroup
    tag="ul"
  >
    <InfoItem
      label="Updated On"
    >
      <span
        className="mr-2"
      >
        Unknown
      </span>
      <Button
        className="details-refresh"
        color="primary"
        onClick={[Function]}
        outline={true}
        size="sm"
        tag="button"
      >
        Refresh
      </Button>
    </InfoItem>
  </ListGroup>
</div>
`;
